File: packages/xyd-openapi2opensdk/src/security.ts
import type { OpenAPIV3 } from 'openapi-types';
import type { SdkSecurity } from '@xyd-js/opensdk-core';

import { screamingSnakeCase } from './naming';

/** Map one OpenAPI securityScheme object into the OpenSDK security shape. */
export function mapSecurityScheme(value: unknown, schemeName?: string, envVar?: string): SdkSecurity | undefined {
  const scheme = value as OpenAPIV3.SecuritySchemeObject;
  if (!scheme || typeof scheme !== 'object' || !('type' in scheme)) return undefined;

  const entry: SdkSecurity = { type: scheme.type, kind: 'other' };
  if (schemeName) entry.schemeName = schemeName;
  if (envVar) entry.envVar = envVar;

  if (scheme.type === 'http') {
    entry.scheme = scheme.scheme;
    if ('bearerFormat' in scheme && scheme.bearerFormat) entry.bearerFormat = scheme.bearerFormat;
    const s = (scheme.scheme || '').toLowerCase();
    entry.kind = s === 'basic' ? 'basic' : 'bearer';
  } else if (scheme.type === 'apiKey') {
    entry.in = scheme.in;
    entry.name = scheme.name;
    entry.kind = scheme.in === 'query' ? 'apiKey-query' : scheme.in === 'cookie' ? 'apiKey-cookie' : 'apiKey-header';
  }

  return entry;
}

/**
 * Map the document's `components.securitySchemes` to the OpenSDK security shape,
 * attaching a best-guess env var name a generated SDK reads the credential from.
 */
export function securitySchemes(doc: OpenAPIV3.Document, sdkName: string, authEnvVar?: string): SdkSecurity[] {
  const schemes = doc.components?.securitySchemes;
  if (!schemes) return [];

  const defaultEnv = authEnvVar || `${screamingSnakeCase(sdkName)}_API_KEY`;
  const out: SdkSecurity[] = [];

  for (const [schemeName, value] of Object.entries(schemes)) {
    const entry = mapSecurityScheme(value, schemeName, defaultEnv);
    if (entry) out.push(entry);
  }

  return out;
}

/**
 * Map a per-operation `security` requirement list through
 * `components.securitySchemes` into the OpenSDK shape. An empty input (an
 * explicitly public endpoint) maps to an empty list — distinct from "inherit
 * the document default", which is expressed by not setting Method.security.
 */
export function securityRequirements(
  doc: OpenAPIV3.Document,
  requirements: OpenAPIV3.SecurityRequirementObject[],
): SdkSecurity[] {
  const schemes = (doc.components?.securitySchemes ?? {}) as Record<string, unknown>;
  const out: SdkSecurity[] = [];
  for (const requirement of requirements) {
    for (const schemeName of Object.keys(requirement)) {
      const entry = mapSecurityScheme(schemes[schemeName], schemeName);
      if (entry) out.push(entry);
    }
  }
  return out;
}

File: crates/xyd_openapi2opencli/src/security.rs
//! securitySchemesToXOpenApi — port of security.ts.

use serde_json::Value;

use crate::jsrt::{js_object_keys, screaming_snake_case};
use crate::model::Security;

pub fn security_schemes_to_x_openapi(
    doc: &Value,
    cli_name: &str,
    auth_env_var: Option<&str>,
) -> Vec<Security> {
    let Some(schemes) = doc
        .get("components")
        .and_then(|c| c.get("securitySchemes"))
        .and_then(|s| s.as_object())
    else {
        return vec![];
    };

    let default_env = auth_env_var
        .map(|s| s.to_string())
        .unwrap_or_else(|| format!("{}_API_KEY", screaming_snake_case(cli_name)));

    let mut out = Vec::new();
    // JS `for (const value of Object.values(schemes))` — value order.
    for key in js_object_keys(schemes) {
        let scheme = &schemes[key];
        let Some(scheme_type) = scheme.get("type") else {
            continue;
        };

        let mut entry = Security {
            scheme_type: scheme_type.clone(),
            kind: "other".to_string(),
            env_var: default_env.clone(),
            scheme: None,
            bearer_format: None,
            location: None,
            name: None,
        };

        match scheme_type.as_str() {
            Some("http") => {
                entry.scheme = scheme.get("scheme").cloned();
                if let Some(bf) = scheme.get("bearerFormat") {
                    let truthy = crate::jsrt::truthy(Some(bf));
                    if truthy {
                        entry.bearer_format = Some(bf.clone());
                    }
                }
                let s = scheme
                    .get("scheme")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_lowercase();
                entry.kind = if s == "basic" { "basic" } else { "bearer" }.to_string();
            }
            Some("apiKey") => {
                entry.location = scheme.get("in").cloned();
                entry.name = scheme.get("name").cloned();
                entry.kind = match scheme.get("in").and_then(|v| v.as_str()) {
                    Some("query") => "apiKey-query",
                    Some("cookie") => "apiKey-cookie",
                    _ => "apiKey-header",
                }
                .to_string();
            }
            _ => {}
        }

        out.push(entry);
    }
    out
}
